File: src/Watari.WebView/native/macos/webview_bridge.m
#import <Cocoa/Cocoa.h>
#include <Foundation/Foundation.h>
#import <WebKit/WebKit.h>
#import <objc/runtime.h>
#include <stdlib.h>

@interface WebViewConsoleLogger : NSObject <WKScriptMessageHandler>
@property(nonatomic) void (*consoleCallback)
    (const char *level, const char *message);
@end

@implementation WebViewConsoleLogger
- (void)userContentController:(WKUserContentController *)userContentController
      didReceiveScriptMessage:(WKScriptMessage *)message {
  if ([message.name isEqualToString:@"consoleLog"]) {
    NSDictionary *dict = (NSDictionary *)message.body;
    NSString *level = dict[@"level"];
    NSString *msg = dict[@"message"];
    if (self.consoleCallback) {
      self.consoleCallback([level UTF8String], [msg UTF8String]);
    }
  }
}
@end

@interface WebViewDropHandler : NSObject <WKScriptMessageHandler>
@property(nonatomic, assign) WKWebView *webView;
@property(nonatomic, strong) NSString *callbackId;
@property(nonatomic, strong) NSString *selector;
@property(nonatomic, strong) NSString *allowedExtensions;
@end

@implementation WebViewDropHandler
- (void)userContentController:(WKUserContentController *)userContentController
      didReceiveScriptMessage:(WKScriptMessage *)message {
  if ([message.name isEqualToString:@"setDropZone"]) {
    NSDictionary *dict = (NSDictionary *)message.body;
    self.callbackId = dict[@"callbackId"];
    self.selector = dict[@"element"];
    self.allowedExtensions = dict[@"allowedExtensions"];
  } else if ([message.name isEqualToString:@"removeDropZone"]) {
    self.callbackId = nil;
    self.selector = nil;
    self.allowedExtensions = nil;
  }
}
@end

@interface WKWebView (DropSupport) <NSDraggingDestination>
@property(nonatomic) WebViewDropHandler *dropHandler;
@end

@implementation WKWebView (DropSupport)
- (void)setDropHandler:(WebViewDropHandler *)handler {
  objc_setAssociatedObject(self, @selector(dropHandler), handler,
                           OBJC_ASSOCIATION_RETAIN_NONATOMIC);
}
- (WebViewDropHandler *)dropHandler {
  return objc_getAssociatedObject(self, @selector(dropHandler));
}
- (NSDragOperation)draggingEntered:(id<NSDraggingInfo>)sender {
  WebViewDropHandler *handler = self.dropHandler;
  if (!handler || !handler.selector)
    return NSDragOperationNone;
  NSPasteboard *pasteboard = [sender draggingPasteboard];
  NSArray *fileURLs =
      [pasteboard canReadObjectForClasses:@[ [NSURL class] ]
                                  options:@{
                                    NSPasteboardURLReadingFileURLsOnlyKey : @YES
                                  }]
          ? [pasteboard
                readObjectsForClasses:@[ [NSURL class] ]
                              options:@{
                                NSPasteboardURLReadingFileURLsOnlyKey : @YES
                              }]
          : @[];
  BOOL hasFiles = fileURLs.count > 0;
  if (!hasFiles)
    return NSDragOperationNone;

  NSPoint location = [sender draggingLocation];
  NSPoint pointInView = [self convertPoint:location fromView:nil];
  NSString *js = [NSString
      stringWithFormat:@"watari._checkDropZone('%@', %f, %f)", handler.selector,
                       pointInView.x, pointInView.y];
  [self evaluateJavaScript:js
         completionHandler:^(id result, NSError *error) {
           if (error) {
             NSLog(@"[WebView_Drag] JS error: %@", error);
           }
           BOOL isOver =
               [result isKindOfClass:[NSNumber class]] && [result boolValue];
           BOOL allowed = YES;
           if (handler.allowedExtensions &&
               handler.allowedExtensions.length > 0) {
             NSArray *allowedExts =
                 [handler.allowedExtensions componentsSeparatedByString:@","];
             for (NSURL *url in fileURLs) {
               NSString *path = [url path];
               NSString *ext = [path pathExtension];
               if (![allowedExts containsObject:ext]) {
                 allowed = NO;
                 break;
               }
             }
           }
           NSString *js2 = [NSString
               stringWithFormat:@"watari._updateDropZoneClass('%@', %d, %d)",
                                handler.selector, isOver ? 1 : 0,
                                allowed ? 1 : 0];
           [self evaluateJavaScript:js2 completionHandler:nil];
         }];
  return NSDragOperationCopy;
}
- (NSDragOperation)draggingUpdated:(id<NSDraggingInfo>)sender {
  return [self draggingEntered:sender];
}
- (void)draggingExited:(id<NSDraggingInfo>)sender {
  WebViewDropHandler *handler = self.dropHandler;
  if (handler && handler.selector) {
    NSString *clearJs = [NSString
        stringWithFormat:@"watari._clearDropZoneClass('%@')", handler.selector];
    [self evaluateJavaScript:clearJs completionHandler:nil];
  }
}
- (BOOL)prepareForDragOperation:(id<NSDraggingInfo>)sender {
  return YES;
}
- (BOOL)performDragOperation:(id<NSDraggingInfo>)sender {
  WebViewDropHandler *handler = self.dropHandler;
  NSString *clearJs = [NSString
      stringWithFormat:@"watari._clearDropZoneClass('%@')", handler.selector];
  [self evaluateJavaScript:clearJs completionHandler:nil];
  if (!handler.callbackId)
    return NO;
  NSPasteboard *pasteboard = [sender draggingPasteboard];
  NSArray *fileURLs = [pasteboard
      readObjectsForClasses:@[ [NSURL class] ]
                    options:@{NSPasteboardURLReadingFileURLsOnlyKey : @YES}];
  NSMutableArray *paths = [NSMutableArray array];
  for (NSURL *url in fileURLs) {
    if ([url isFileURL]) {
      [paths addObject:[url path]];
    }
  }
  if (paths.count == 0)
    return NO;

  NSPoint location = [sender draggingLocation];
  NSPoint pointInView = [self convertPoint:location fromView:nil];
  NSString *js = [NSString
      stringWithFormat:@"watari._checkDropZone('%@', %f, %f)", handler.selector,
                       pointInView.x, pointInView.y];
  [self evaluateJavaScript:js
         completionHandler:^(id result, NSError *error) {
           if (error) {
             NSLog(@"[WebView_Drag] JS error: %@", error);
           }
           BOOL isOver =
               [result isKindOfClass:[NSNumber class]] && [result boolValue];
           if (isOver) {
             NSError *error;
             NSData *jsonData = [NSJSONSerialization dataWithJSONObject:paths
                                                                options:0
                                                                  error:&error];
             if (jsonData) {
               NSString *jsonString =
                   [[NSString alloc] initWithData:jsonData
                                         encoding:NSUTF8StringEncoding];
               NSString *js2 =
                   [NSString stringWithFormat:@"watari.callbacks[%@] (%@)",
                                              handler.callbackId, jsonString];
               [self evaluateJavaScript:js2 completionHandler:nil];
             }
           }
         }];
  return YES;
}
@end

__attribute__((visibility("default"))) CFTypeRef
WebView_Create(void (*callback)(const char *, const char *)) {
  @autoreleasepool {
    NSLog(@"[WebView_Create] creating WKWebView");
    WKWebViewConfiguration *config = [[WKWebViewConfiguration alloc] init];

    // Set up console.log redirection
    WKUserContentController *userContentController =
        [[WKUserContentController alloc] init];
    WebViewConsoleLogger *logger = [[WebViewConsoleLogger alloc] init];
    logger.consoleCallback = callback;
    [userContentController addScriptMessageHandler:logger name:@"consoleLog"];

    // Set up drop zone handler
    WebViewDropHandler *dropHandler = [[WebViewDropHandler alloc] init];
    [userContentController addScriptMessageHandler:dropHandler
                                              name:@"setDropZone"];
    [userContentController addScriptMessageHandler:dropHandler
                                              name:@"removeDropZone"];

    // Inject script to override console methods
    NSString *consoleOverrideScript =
        @"var levels = ['log', 'error', 'warn', 'info', 'debug']; "
        @"levels.forEach(function(level) { console[level] = function(...args) "
        @"{ window.webkit.messageHandlers.consoleLog.postMessage({level: "
        @"level, message: args.join(' ')}); }; });";
    WKUserScript *userScript = [[WKUserScript alloc]
          initWithSource:consoleOverrideScript
           injectionTime:WKUserScriptInjectionTimeAtDocumentStart
        forMainFrameOnly:YES];
    [userContentController addUserScript:userScript];

    [config setUserContentController:userContentController];

    WKWebView *created =
        [[WKWebView alloc] initWithFrame:NSMakeRect(0, 0, 1024, 768)
                           configuration:config];
    NSLog(@"[WebView_Create] created WKWebView %p", created);

    // Set up drop handler and register for drag types
    dropHandler.webView = created;
    [created setDropHandler:dropHandler];
    [created registerForDraggedTypes:@[ NSPasteboardTypeFileURL ]];

    return CFBridgingRetain(created);
  }
}

__attribute__((visibility("default"))) void WebView_Navigate(void *viewHandle,
                                                             const char *url) {
  @autoreleasepool {
    WKWebView *webView = (__bridge WKWebView *)viewHandle;
    NSString *nsurl = [NSString stringWithUTF8String:url];
    NSURL *u = [NSURL URLWithString:nsurl];
    NSLog(@"[WebView_Navigate] created NSURL: %@", u);
    NSURLRequest *req = [NSURLRequest requestWithURL:u];
    NSLog(@"[WebView_Navigate] navigating WKWebView %p to URL: %@", webView,
          nsurl);
    [webView loadRequest:req];
  }
}

__attribute__((visibility("default"))) void WebView_Eval(void *viewHandle,
                                                         const char *js) {
  @autoreleasepool {
    WKWebView *webView = (__bridge WKWebView *)viewHandle;
    NSString *jsString = [NSString stringWithUTF8String:js];
    NSLog(@"[WebView_Eval] evaluating JavaScript on WKWebView %p: %@", webView,
          jsString);
    [webView
        evaluateJavaScript:jsString
         completionHandler:^(id result, NSError *error) {
           if (error) {
             NSLog(@"[WebView_Eval] JavaScript evaluation error: %@", error);
           } else {
             NSLog(@"[WebView_Eval] JavaScript evaluation result: %@", result);
           }
         }];
  }
}

__attribute__((visibility("default"))) void WebView_Destroy(void *viewHandle) {
  @autoreleasepool {
    if (!viewHandle)
      return;
    WKWebView *webView = (__bridge WKWebView *)viewHandle;
    NSLog(@"[WebView_Destroy] destroying WKWebView %p", webView);
    // Remove from superview if attached
    [webView removeFromSuperview];
    // Release the webView
    webView = nil;
    NSLog(@"[WebView_Destroy] destroyed WKWebView");
  }
}

__attribute__((visibility("default"))) void
WebView_AddUserScript(void *viewHandle, const char *scriptSource,
                      int injectionTime, bool forMainFrameOnly) {
  @autoreleasepool {
    WKWebView *webView = (__bridge WKWebView *)viewHandle;
    NSString *source = [NSString stringWithUTF8String:scriptSource];
    WKUserScriptInjectionTime time = (WKUserScriptInjectionTime)injectionTime;
    WKUserScript *userScript =
        [[WKUserScript alloc] initWithSource:source
                               injectionTime:time
                            forMainFrameOnly:forMainFrameOnly];
    [webView.configuration.userContentController addUserScript:userScript];
  }
}

File: src/Watari.WebView/native/macos/webview_bridge.swift
import Cocoa
import WebKit
import UniformTypeIdentifiers

func commonJSCompletionHandler(webView: WKWebView, callbackId: String? = nil) -> (Any?, Error?) -> Void {
    return { result, error in
        if let error = error {
            print("[WebView] JavaScript evaluation error: \(error.localizedDescription)")
            let nsError = error as NSError 
            print("[WebView] Error domain: \(nsError.domain), code: \(nsError.code)")
            print("[WebView] Error userInfo: \(nsError.userInfo)")
            if callbackId != nil {
                let escapedError = error.localizedDescription.replacingOccurrences(of: "'", with: "\\'")
                let errorJs = "watari.callbackError('\(callbackId!)', '\(escapedError)')"
                webView.evaluateJavaScript(errorJs, completionHandler: nil)
            }
        }
    }
}

class WebViewConsoleLogger: NSObject, WKScriptMessageHandler {
    var consoleCallback: (@convention(c) (UnsafePointer<CChar>?, UnsafePointer<CChar>?) -> Void)?

    func userContentController(_ userContentController: WKUserContentController, didReceive message: WKScriptMessage) {
        if message.name == "consoleLog" {
            if let dict = message.body as? [String: Any],
               let level = dict["level"] as? String,
               let msg = dict["message"] as? String {
                consoleCallback?(level, msg)
            }
        }
    }
}

class WebViewDropHandler: NSObject, WKScriptMessageHandler {
    weak var webView: WKWebView?
    var callbackId: String?
    var id: String?
    var allowedExtensions: String?

    func userContentController(_ userContentController: WKUserContentController, didReceive message: WKScriptMessage) {
        if message.name == "setDropZone" {
            if let dict = message.body as? [String: Any] {
                callbackId = dict["callbackId"] as? String
                id = dict["element"] as? String
                allowedExtensions = dict["allowedExtensions"] as? String
            }
        } else if message.name == "removeDropZone" {
            callbackId = nil
            id = nil
            allowedExtensions = nil
        }
    }
}

class WebViewFileDialogHandler: NSObject, WKScriptMessageHandler {
    weak var webView: WKWebView?

    func userContentController(_ userContentController: WKUserContentController, didReceive message: WKScriptMessage) {
        if message.name == "openFileDialog" {
            if let dict = message.body as? [String: Any],
               let callbackId = dict["callbackId"] as? String,
               let allowedExtensions = dict["allowedExtensions"] as? String {
                openFileDialog(callbackId: callbackId, allowedExtensions: allowedExtensions)
            }
        }
    }

    func openFileDialog(callbackId: String, allowedExtensions: String) {
        let extensions = allowedExtensions.split(separator: ",").map { String($0).trimmingCharacters(in: .whitespaces) }
        
        let panel = NSOpenPanel()
        panel.canChooseFiles = true
        panel.canChooseDirectories = false
        panel.allowsMultipleSelection = false
        if #available(macOS 12.0, *) {
            panel.allowedContentTypes = extensions.compactMap { UTType(filenameExtension: $0) }
        } else {
            panel.allowedFileTypes = extensions
        }
        
        let response = panel.runModal()
        var result: String? = nil
        if response == .OK, let url = panel.url {
            result = url.path
        }
        print("[WebViewFileDialogHandler] selected file: \(result ?? "none")")
        let js: String
        if let result = result {
            let escapedResult = result.replacingOccurrences(of: "'", with: "\\'")
            js = "watari.callbacks['\(callbackId)']('\(escapedResult)')"
        } else {
            js = "watari.callbacks['\(callbackId)'](null)"
        }
        webView?.evaluateJavaScript(js, completionHandler: commonJSCompletionHandler(webView: webView!, callbackId: callbackId))
    }
}

class MyWKWebView: WKWebView {
    var dropHandler: WebViewDropHandler?

    override func draggingEntered(_ sender: NSDraggingInfo) -> NSDragOperation {
        guard let handler = dropHandler, handler.id != nil else { return [] }
        let pasteboard = sender.draggingPasteboard
        let fileURLs = pasteboard.readObjects(forClasses: [NSURL.self], options: [.urlReadingFileURLsOnly: true]) as? [URL] ?? []
        guard !fileURLs.isEmpty else { return [] }
        var paths = [String]()
        for url in fileURLs {
            if url.isFileURL {
                paths.append(url.path)
            }
        }
        guard !paths.isEmpty else { return [] }
        guard let jsonData = try? JSONSerialization.data(withJSONObject: paths, options: []),
              let jsonString = String(data: jsonData, encoding: .utf8) else { return [] }
        let location = sender.draggingLocation
        let pointInView = convert(location, from: nil)
        let js = "watari._checkAndValidateDropZone('\(handler.id!)', \(pointInView.x), \(pointInView.y), '\(handler.allowedExtensions ?? "")', '\(jsonString)')"
        evaluateJavaScript(js, completionHandler: commonJSCompletionHandler(webView: self))
        return .copy
    }

    override func draggingUpdated(_ sender: NSDraggingInfo) -> NSDragOperation {
        draggingEntered(sender)
    }

    override func draggingExited(_ sender: NSDraggingInfo?) {
        guard let handler = dropHandler, handler.id != nil else { return }
        let clearJs = "watari._clearDropZoneClass('\(handler.id!)')"
        evaluateJavaScript(clearJs, completionHandler: commonJSCompletionHandler(webView: self))
    }

    override func prepareForDragOperation(_ sender: NSDraggingInfo) -> Bool {
        true
    }

    override func performDragOperation(_ sender: NSDraggingInfo) -> Bool {
        guard let handler = dropHandler, let callbackId = handler.callbackId else { return false }
        let clearJs = "watari._clearDropZoneClass('\(handler.id!)')"
        evaluateJavaScript(clearJs, completionHandler: commonJSCompletionHandler(webView: self))
        let pasteboard = sender.draggingPasteboard
        let fileURLs = pasteboard.readObjects(forClasses: [NSURL.self], options: [.urlReadingFileURLsOnly: true]) as? [URL] ?? []
        var paths = [String]()
        for url in fileURLs {
            if url.isFileURL {
                paths.append(url.path)
            }
        }
        guard !paths.isEmpty,
              let jsonData = try? JSONSerialization.data(withJSONObject: paths, options: []),
              let jsonString = String(data: jsonData, encoding: .utf8) else { return false }
        let location = sender.draggingLocation
        let pointInView = convert(location, from: nil)
        let js = "watari._handleDrop('\(handler.id!)', \(pointInView.x), \(pointInView.y), '\(jsonString)', '\(callbackId)')"
        evaluateJavaScript(js, completionHandler: commonJSCompletionHandler(webView: self, callbackId: callbackId))
        return true
    }
}

@_cdecl("WebView_Create")
public func WebView_Create(_ callback: (@convention(c) (UnsafePointer<CChar>?, UnsafePointer<CChar>?) -> Void)?) -> CFTypeRef? {
    print("[WebView_Create] creating WKWebView")
    let config = WKWebViewConfiguration()

    // Set up console.log redirection
    let userContentController = WKUserContentController()
    let logger = WebViewConsoleLogger()
    logger.consoleCallback = callback
    userContentController.add(logger, name: "consoleLog")

    // Set up drop zone handler
    let dropHandler = WebViewDropHandler()
    userContentController.add(dropHandler, name: "setDropZone")
    userContentController.add(dropHandler, name: "removeDropZone")

    // Set up file dialog handler
    let fileDialogHandler = WebViewFileDialogHandler()
    userContentController.add(fileDialogHandler, name: "openFileDialog")
    config.userContentController = userContentController

    let created = MyWKWebView(frame: NSRect(x: 0, y: 0, width: 1024, height: 768), configuration: config)
    print("[WebView_Create] created WKWebView \(created)")

    // Set up drop handler and register for drag types
    dropHandler.webView = created
    created.dropHandler = dropHandler
    created.registerForDraggedTypes([.fileURL])

    // Set webView for file dialog handler
    fileDialogHandler.webView = created

    return CFBridgingRetain(created)
}

@_cdecl("WebView_SetEnableDevTools")
public func WebView_SetEnableDevTools(_ viewHandle: UnsafeMutableRawPointer?, _ enable: Bool) {
    guard let viewHandle = viewHandle else { return }
    let webView = Unmanaged<WKWebView>.fromOpaque(viewHandle).takeUnretainedValue()
    webView.configuration.preferences.setValue(enable, forKey: "developerExtrasEnabled")
}

@_cdecl("WebView_Navigate")
public func WebView_Navigate(_ viewHandle: UnsafeMutableRawPointer?, _ url: UnsafePointer<CChar>?) {
    guard let viewHandle = viewHandle, let url = url else { return }
    let webView = Unmanaged<WKWebView>.fromOpaque(viewHandle).takeUnretainedValue()
    let nsurl = String(cString: url)
    guard let u = URL(string: nsurl) else { return }
    print("[WebView_Navigate] created URL: \(u)")
    let req = URLRequest(url: u)
    print("[WebView_Navigate] navigating WKWebView \(webView) to URL: \(nsurl)")
    webView.load(req)
}

@_cdecl("WebView_Eval")
public func WebView_Eval(_ viewHandle: UnsafeMutableRawPointer?, _ js: UnsafePointer<CChar>?) {
    guard let viewHandle = viewHandle, let js = js else { return }
    let webView = Unmanaged<WKWebView>.fromOpaque(viewHandle).takeUnretainedValue()
    let jsString = String(cString: js)
    print("[WebView_Eval] evaluating JavaScript on WKWebView \(webView): \(jsString)")
    webView.evaluateJavaScript(jsString, completionHandler: commonJSCompletionHandler(webView: webView))
}

@_cdecl("WebView_Destroy")
public func WebView_Destroy(_ viewHandle: UnsafeMutableRawPointer?) {
    guard let viewHandle = viewHandle else { return }
    let webView = Unmanaged<WKWebView>.fromOpaque(viewHandle).takeUnretainedValue()
    webView.removeFromSuperview()
}

@_cdecl("WebView_AddUserScript")
public func WebView_AddUserScript(_ viewHandle: UnsafeMutableRawPointer?, _ scriptSource: UnsafePointer<CChar>?, _ injectionTime: Int32, _ forMainFrameOnly: Bool) {
    guard let viewHandle = viewHandle, let scriptSource = scriptSource else { return }
    let webView = Unmanaged<WKWebView>.fromOpaque(viewHandle).takeUnretainedValue()
    let source = String(cString: scriptSource)
    let time = WKUserScriptInjectionTime(rawValue: Int(injectionTime)) ?? .atDocumentStart
    let userScript = WKUserScript(source: source, injectionTime: time, forMainFrameOnly: forMainFrameOnly)
    webView.configuration.userContentController.addUserScript(userScript)
}